README.md
SWTableViewCell
===============

[![Build Status](https://travis-ci.org/addoshi/SWTableViewCell.svg?branch=master)](https://travis-ci.org/addoshi/SWTableViewCell)
[![Coverage Status](https://coveralls.io/repos/addoshi/SWTableViewCell/badge.svg)](https://coveralls.io/r/addoshi/SWTableViewCell)

<p align="center"><img src="http://i.imgur.com/njKCjK8.gif"/></p>

An easy-to-use UITableViewCell subclass that implements a swipeable content view which exposes utility buttons (similar to iOS 7 Mail Application)

##Usage
In your Podfile:
<pre>pod 'SWTableViewCell', '~> 0.3.7'</pre>

Or just clone this repo and manually add source to project

##Functionality
###Right Utility Buttons
Utility buttons that become visible on the right side of the Table View Cell when the user swipes left. This behavior is similar to that seen in the iOS apps Mail and Reminders.

<p align="center"><img src="http://i.imgur.com/gDZFRpr.gif"/></p>

###Left Utility Buttons
Utility buttons that become visible on the left side of the Table View Cell when the user swipes right. 

<p align="center"><img src="http://i.imgur.com/qt6aISz.gif"/></p>

###Features
* Dynamic utility button scalling. As you add more buttons to a cell, the other buttons on that side get smaller to make room
* Smart selection: The cell will pick up touch events and either scroll the cell back to center or fire the delegate method `- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath` 
<p align="center"><img src="http://i.imgur.com/TYGx9h8.gif"/></p>
So the cell will not be considered selected when the user touches the cell while utility buttons are visible, instead the cell will slide back into place (same as iOS 7 Mail App functionality)
* Create utilty buttons with either a title or an icon along with a RGB color
* Tested on iOS 6.1 and above, including iOS 7

##Usage

###Standard Table View Cells

In your `tableView:cellForRowAtIndexPath:` method you set up the SWTableView cell and add an arbitrary amount of utility buttons to it using the included `NSMutableArray+SWUtilityButtons` category.

```objc
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"Cell";
    
    SWTableViewCell *cell = (SWTableViewCell *)[tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    if (cell == nil) {
        cell = [[SWTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
        cell.leftUtilityButtons = [self leftButtons];
        cell.rightUtilityButtons = [self rightButtons];
        cell.delegate = self;
    }
    
    NSDate *dateObject = _testArray[indexPath.row];
    cell.textLabel.text = [dateObject description];
    cell.detailTextLabel.text = @"Some detail text";
    
    return cell;
}

- (NSArray *)rightButtons
{
    NSMutableArray *rightUtilityButtons = [NSMutableArray new];
    [rightUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:0.78f green:0.78f blue:0.8f alpha:1.0]
                                                title:@"More"];
    [rightUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:1.0f green:0.231f blue:0.188 alpha:1.0f]
                                                title:@"Delete"];

    return rightUtilityButtons;
}

- (NSArray *)leftButtons
{
    NSMutableArray *leftUtilityButtons = [NSMutableArray new];
    
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:0.07 green:0.75f blue:0.16f alpha:1.0]
                                                icon:[UIImage imageNamed:@"check.png"]];
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:1.0f green:1.0f blue:0.35f alpha:1.0]
                                                icon:[UIImage imageNamed:@"clock.png"]];
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:1.0f green:0.231f blue:0.188f alpha:1.0]
                                                icon:[UIImage imageNamed:@"cross.png"]];
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:0.55f green:0.27f blue:0.07f alpha:1.0]
                                                icon:[UIImage imageNamed:@"list.png"]];
    
    return leftUtilityButtons;
}
```

###Custom Table View Cells

Thanks to [Matt Bowman](https://github.com/MattCBowman) you can now create custom table view cells using Interface Builder that have the capabilities of an SWTableViewCell

The first step is to design your cell either in a standalone nib or inside of a
table view using prototype cells. Make sure to set the custom class on the 
cell in interface builder to the subclass you made for it:

<p align="center"><img src="http://i.imgur.com/mfrT1Ex.png"/></p>

Then set the cell reuse identifier:

<p align="center"><img src="http://i.imgur.com/dlmArZ1.png"/></p>

When writing your custom table view cell's code, make sure your cell is a
subclass of SWTableViewCell:

```objc

#import <SWTableViewCell.h>

@interface MyCustomTableViewCell : SWTableViewCell

@property (weak, nonatomic) UILabel *customLabel;
@property (weak, nonatomic) UIImageView *customImageView;

@end

```

If you are using a separate nib and not a prototype cell, you'll need to be sure to register the nib in your table view:

```objc

- (void)viewDidLoad
{
    [super viewDidLoad];

    [self.tableView registerNib:[UINib nibWithNibName:@"MyCustomTableViewCellNibFileName" bundle:nil] forCellReuseIdentifier:@"MyCustomCell"];
}

```

Then, in the `tableView:cellForRowAtIndexPath:` method of your `UITableViewDataSource` (usually your view controller), initialize your custom cell:

```objc
- (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath
{
    static NSString *cellIdentifier = @"MyCustomCell";
    
    MyCustomTableViewCell *cell = (MyCustomTableViewCell *)[tableView dequeueReusableCellWithIdentifier:cellIdentifier 
                                                                                           forIndexPath:indexPath];

    cell.leftUtilityButtons = [self leftButtons];
    cell.rightUtilityButtons = [self rightButtons];
    cell.delegate = self;
    
    cell.customLabel.text = @"Some Text";
    cell.customImageView.image = [UIImage imageNamed:@"MyAwesomeTableCellImage"];
    [cell setCellHeight:cell.frame.size.height];
    return cell;
}
```

###Delegate

The delegate `SWTableViewCellDelegate` is used by the developer to find out which button was pressed. There are five methods:

```objc
// click event on left utility button
- (void)swipeableTableViewCell:(SWTableViewCell *)cell didTriggerLeftUtilityButtonWithIndex:(NSInteger)index;

// click event on right utility button
- (void)swipeableTableViewCell:(SWTableViewCell *)cell didTriggerRightUtilityButtonWithIndex:(NSInteger)index;

// utility button open/close event
- (void)swipeableTableViewCell:(SWTableViewCell *)cell scrollingToState:(SWCellState)state;

// prevent multiple cells from showing utilty buttons simultaneously
- (BOOL)swipeableTableViewCellShouldHideUtilityButtonsOnSwipe:(SWTableViewCell *)cell;

// prevent cell(s) from displaying left/right utility buttons
- (BOOL)swipeableTableViewCell:(SWTableViewCell *)cell canSwipeToState:(SWCellState)state;
```

The index signifies which utility button the user pressed, for each side the button indices are ordered from right to left 0...n

####Example

```objc
#pragma mark - SWTableViewDelegate

- (void)swipeableTableViewCell:(SWTableViewCell *)cell didTriggerLeftUtilityButtonWithIndex:(NSInteger)index {
    switch (index) {
        case 0:
            NSLog(@"check button was pressed");
            break;
        case 1:
            NSLog(@"clock button was pressed");
            break;
        case 2:
            NSLog(@"cross button was pressed");
            break;
        case 3:
            NSLog(@"list button was pressed");
        default:
            break;
    }
}

- (void)swipeableTableViewCell:(SWTableViewCell *)cell didTriggerRightUtilityButtonWithIndex:(NSInteger)index {
    switch (index) {
        case 0:
            NSLog(@"More button was pressed");
            break;
        case 1:
        {
            // Delete button was pressed
            NSIndexPath *cellIndexPath = [self.tableView indexPathForCell:cell];
            
            [_testArray removeObjectAtIndex:cellIndexPath.row];
            [self.tableView deleteRowsAtIndexPaths:@[cellIndexPath] 
                    withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
        }
        default:
            break;
    }
}
```

(This is all code from the included example project)

###Gotchas

#### Seperator Insets
* If you have left utility button on iOS 7, I recommend changing your Table View's seperatorInset so the seperator stretches the length of the screen
<pre> tableView.separatorInset = UIEdgeInsetsMake(0, 0, 0, 0); </pre>


##Contributing
Use [Github issues](https://github.com/cewendel/SWTableViewCell/issues) to track bugs and feature requests.

##Contact

Chris Wendel

- http://twitter.com/CEWendel

## Licence

MIT 







CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/CEWendel/SWTableViewCell/tree/HEAD)

[Full Changelog](https://github.com/CEWendel/SWTableViewCell/compare/0.3.7...HEAD)

**Closed issues:**

- Mistake in Readme [\#315](https://github.com/CEWendel/SWTableViewCell/issues/315)

- Background color of 2 buttons overlap creating a  third color [\#294](https://github.com/CEWendel/SWTableViewCell/issues/294)

- Space between 2 buttons [\#293](https://github.com/CEWendel/SWTableViewCell/issues/293)

- As a customer, I'd like to use the custom UIView instead of the Array of buttons. [\#279](https://github.com/CEWendel/SWTableViewCell/issues/279)

- Expected type in NSMutableArray + SWutility.h button [\#278](https://github.com/CEWendel/SWTableViewCell/issues/278)

- Vertically-aligned buttons [\#277](https://github.com/CEWendel/SWTableViewCell/issues/277)

- Similar effect in Android [\#276](https://github.com/CEWendel/SWTableViewCell/issues/276)

- Swiping Left or Right Removes Cell Accessory Type [\#274](https://github.com/CEWendel/SWTableViewCell/issues/274)

- iPhone 6 Bug [\#267](https://github.com/CEWendel/SWTableViewCell/issues/267)

- send msg to zombie object [\#266](https://github.com/CEWendel/SWTableViewCell/issues/266)

- iPhone 6 Plus performance issue [\#246](https://github.com/CEWendel/SWTableViewCell/issues/246)

- Big BUG in IOS 8! \[self.tableview reloadData\] opens previously closed utility buttons. [\#243](https://github.com/CEWendel/SWTableViewCell/issues/243)

- Ability to add custom uiview as utility buttons [\#159](https://github.com/CEWendel/SWTableViewCell/issues/159)

- With right utility buttons, it is possible to get to kCellStateLeft on close. [\#126](https://github.com/CEWendel/SWTableViewCell/issues/126)

**Merged pull requests:**

- Added a delegate that exposes the scroll state of the cell.  [\#296](https://github.com/CEWendel/SWTableViewCell/pull/296) ([aluong-yammer](https://github.com/aluong-yammer))

- it prevents from sending scrollview delegate method to deallocated cell [\#286](https://github.com/CEWendel/SWTableViewCell/pull/286) ([adamlipka](https://github.com/adamlipka))

## [0.3.7](https://github.com/CEWendel/SWTableViewCell/tree/0.3.7) (2014-12-19)

[Full Changelog](https://github.com/CEWendel/SWTableViewCell/compare/0.3.6...0.3.7)

**Closed issues:**

- rotating orientation with CUSTOM cells causes left utility buttons to appear and for ALL cells [\#264](https://github.com/CEWendel/SWTableViewCell/issues/264)

- the UITableViewCell dosen't response to prepareforsegue function [\#240](https://github.com/CEWendel/SWTableViewCell/issues/240)

- Frame not respected [\#208](https://github.com/CEWendel/SWTableViewCell/issues/208)

- Rotating from landscape to portrait opens left utility buttons [\#192](https://github.com/CEWendel/SWTableViewCell/issues/192)

- Can't make it work in iOS 7.1 [\#181](https://github.com/CEWendel/SWTableViewCell/issues/181)

- When cell is revealed, quick scrolling of the uitableview crashes app [\#173](https://github.com/CEWendel/SWTableViewCell/issues/173)

- New resued cell already scrolled if starting cell is scrolled [\#168](https://github.com/CEWendel/SWTableViewCell/issues/168)

- Buggy when combining SW and TableView editing delegate methods [\#167](https://github.com/CEWendel/SWTableViewCell/issues/167)

- No centents show on the cell [\#165](https://github.com/CEWendel/SWTableViewCell/issues/165)

- I want to long press on cell, but UILongPressGestureRecognizer work only on button on cell. [\#164](https://github.com/CEWendel/SWTableViewCell/issues/164)

- Why SWTableViewCell not using UIResponder for handling touches?  [\#157](https://github.com/CEWendel/SWTableViewCell/issues/157)

- No extra margin when sectionIndexTitlesForTableView [\#153](https://github.com/CEWendel/SWTableViewCell/issues/153)

- Long app launch time [\#147](https://github.com/CEWendel/SWTableViewCell/issues/147)

- Crash in iOS6.1 [\#139](https://github.com/CEWendel/SWTableViewCell/issues/139)

- Possible to Change cell & scrollView background color while scrolling [\#131](https://github.com/CEWendel/SWTableViewCell/issues/131)

- Vertical Aligned buttons? [\#115](https://github.com/CEWendel/SWTableViewCell/issues/115)

- Support for Vertical stacked buttons  [\#114](https://github.com/CEWendel/SWTableViewCell/issues/114)

- Multi edit Problems [\#110](https://github.com/CEWendel/SWTableViewCell/issues/110)

- Storyboard Segue Not Performed for SWTableViewCell-based Cells [\#109](https://github.com/CEWendel/SWTableViewCell/issues/109)

- canEditRowAtIndexPath not working [\#106](https://github.com/CEWendel/SWTableViewCell/issues/106)

- Swipe gesture too jumpy not 'sticky' enough [\#103](https://github.com/CEWendel/SWTableViewCell/issues/103)

- Button size [\#95](https://github.com/CEWendel/SWTableViewCell/issues/95)

- Disable swipe to right [\#87](https://github.com/CEWendel/SWTableViewCell/issues/87)

- TableViewCell jumps back after full swipe from left to right [\#57](https://github.com/CEWendel/SWTableViewCell/issues/57)

**Merged pull requests:**

- Fix for issue \#264 [\#272](https://github.com/CEWendel/SWTableViewCell/pull/272) ([pbtura](https://github.com/pbtura))

- These are not necessary and impose a huge performance hit during scrolling [\#257](https://github.com/CEWendel/SWTableViewCell/pull/257) ([maciekish](https://github.com/maciekish))

## [0.3.6](https://github.com/CEWendel/SWTableViewCell/tree/0.3.6) (2014-10-26)

[Full Changelog](https://github.com/CEWendel/SWTableViewCell/compare/0.3.5...0.3.6)

**Closed issues:**

- iPhone 6+ Cell State Issue [\#253](https://github.com/CEWendel/SWTableViewCell/issues/253)

- A light tap on cell, no selection color. [\#252](https://github.com/CEWendel/SWTableViewCell/issues/252)

- SWTableViewCell pod is pointing still to release 0.3.0 [\#201](https://github.com/CEWendel/SWTableViewCell/issues/201)

- Update Cocoapod trunk [\#189](https://github.com/CEWendel/SWTableViewCell/issues/189)


SWTableViewCell/ViewController.m
//
//  ViewController.m
//  SWTableViewCell
//
//  Created by Chris Wendel on 9/10/13.
//  Copyright (c) 2013 Chris Wendel. All rights reserved.
//

#import "ViewController.h"
#import "SWTableViewCell.h"
#import "UMTableViewCell.h"

@interface ViewController () {
    NSArray *_sections;
    NSMutableArray *_testArray;
}

@property (nonatomic, weak) IBOutlet UITableView *tableView;
@property (nonatomic) BOOL useCustomCells;
@property (nonatomic, weak) UIRefreshControl *refreshControl;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.leftBarButtonItem = self.editButtonItem;
    self.tableView.rowHeight = 90;
    
    self.navigationItem.title = @"Pull to Toggle Cell Type";
    
    // Setup refresh control for example app
    UIRefreshControl *refreshControl = [[UIRefreshControl alloc] init];
    [refreshControl addTarget:self action:@selector(toggleCells:) forControlEvents:UIControlEventValueChanged];
    refreshControl.tintColor = [UIColor blueColor];

    self.refreshControl = refreshControl;
    
    // If you set the seperator inset on iOS 6 you get a NSInvalidArgumentException...weird
    if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 7) {
        self.tableView.separatorInset = UIEdgeInsetsMake(0, 0, 0, 0); // Makes the horizontal row seperator stretch the entire length of the table view
    }
    
    _sections = [[UILocalizedIndexedCollation currentCollation] sectionIndexTitles];
    
    _testArray = [[NSMutableArray alloc] init];
    
    self.useCustomCells = NO;
    
    for (int i = 0; i < _sections.count; ++i) {
        [_testArray addObject:[NSMutableArray array]];
    }
    
    for (int i = 0; i < 100; ++i) {
        NSString *string = [NSString stringWithFormat:@"%d", i];
        [_testArray[i % _sections.count] addObject:string];
    }
}

#pragma mark UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return _testArray.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [_testArray[section] count];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    NSLog(@"cell selected at index path %ld:%ld", (long)indexPath.section, (long)indexPath.row);
    NSLog(@"selected cell index path is %@", [self.tableView indexPathForSelectedRow]);
    
    if (!tableView.isEditing) {
        [tableView deselectRowAtIndexPath:indexPath animated:YES];
    }
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return _sections[section];
}

// Show index titles

//- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView {
//    return [[UILocalizedIndexedCollation currentCollation] sectionIndexTitles];
//}
//
//- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index {
//    return [[UILocalizedIndexedCollation currentCollation] sectionForSectionIndexTitleAtIndex:index];
//}

#pragma mark - UIRefreshControl Selector

- (void)toggleCells:(UIRefreshControl*)refreshControl
{
    [refreshControl beginRefreshing];
    self.useCustomCells = !self.useCustomCells;
    if (self.useCustomCells)
    {
        self.refreshControl.tintColor = [UIColor yellowColor];
    }
    else
    {
        self.refreshControl.tintColor = [UIColor blueColor];
    }
    [self.tableView reloadData];
    [refreshControl endRefreshing];
}

#pragma mark - UIScrollViewDelegate

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (self.useCustomCells)
    {
        UMTableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"UMCell" forIndexPath:indexPath];
        
        // optionally specify a width that each set of utility buttons will share
        [cell setLeftUtilityButtons:[self leftButtons] WithButtonWidth:32.0f];
        [cell setRightUtilityButtons:[self rightButtons] WithButtonWidth:58.0f];
        cell.delegate = self;
        
        cell.label.text = [NSString stringWithFormat:@"Section: %ld, Seat: %ld", (long)indexPath.section, (long)indexPath.row];
        
        return cell;
    }
    else
    {
        static NSString *cellIdentifier = @"Cell";
        
        SWTableViewCell *cell = (SWTableViewCell *)[tableView dequeueReusableCellWithIdentifier:cellIdentifier];
        
        if (cell == nil) {
            
            cell = [[SWTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
            
            cell.leftUtilityButtons = [self leftButtons];
            cell.rightUtilityButtons = [self rightButtons];
            cell.delegate = self;
        }
        
        NSDate *dateObject = _testArray[indexPath.section][indexPath.row];
        cell.textLabel.text = [dateObject description];
        cell.textLabel.backgroundColor = [UIColor whiteColor];
        cell.detailTextLabel.backgroundColor = [UIColor whiteColor];
        cell.detailTextLabel.text = @"Some detail text";
        
        return cell;
    }
    
}

- (NSArray *)rightButtons
{
    NSMutableArray *rightUtilityButtons = [NSMutableArray new];
    [rightUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:0.78f green:0.78f blue:0.8f alpha:1.0]
                                                title:@"More"];
    [rightUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:1.0f green:0.231f blue:0.188 alpha:1.0f]
                                                title:@"Delete"];
    
    return rightUtilityButtons;
}

- (NSArray *)leftButtons
{
    NSMutableArray *leftUtilityButtons = [NSMutableArray new];
    
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:0.07 green:0.75f blue:0.16f alpha:1.0]
                                                icon:[UIImage imageNamed:@"check.png"]];
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:1.0f green:1.0f blue:0.35f alpha:1.0]
                                                icon:[UIImage imageNamed:@"clock.png"]];
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:1.0f green:0.231f blue:0.188f alpha:1.0]
                                                icon:[UIImage imageNamed:@"cross.png"]];
    [leftUtilityButtons sw_addUtilityButtonWithColor:
     [UIColor colorWithRed:0.55f green:0.27f blue:0.07f alpha:1.0]
                                                icon:[UIImage imageNamed:@"list.png"]];
    
    return leftUtilityButtons;
}

// Set row height on an individual basis

//- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
//    return [self rowHeightForIndexPath:indexPath];
//}
//
//- (CGFloat)rowHeightForIndexPath:(NSIndexPath *)indexPath {
//    return ([indexPath row] * 10) + 60;
//}

- (void)tableView:(UITableView *)tableView willDisplayCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath {
    // Set background color of cell here if you don't want default white
}

#pragma mark - SWTableViewDelegate


- (void)swipeableTableViewCell:(SWTableViewCell *)cell scrollingToState:(SWCellState)state
{
    switch (state) {
        case 0:
            NSLog(@"utility buttons closed");
            break;
        case 1:
            NSLog(@"left utility buttons open");
            break;
        case 2:
            NSLog(@"right utility buttons open");
            break;
        default:
            break;
    }
}

- (void)swipeableTableViewCell:(SWTableViewCell *)cell didTriggerLeftUtilityButtonWithIndex:(NSInteger)index
{
    switch (index) {
        case 0:
            NSLog(@"left button 0 was pressed");
            break;
        case 1:
            NSLog(@"left button 1 was pressed");
            break;
        case 2:
            NSLog(@"left button 2 was pressed");
            break;
        case 3:
            NSLog(@"left btton 3 was pressed");
        default:
            break;
    }
}


- (void)swipeableTableViewCell:(SWTableViewCell *)cell didTriggerRightUtilityButtonWithIndex:(NSInteger)index
{
    switch (index) {
        case 0:
        {
            NSLog(@"More button was pressed");
            UIAlertView *alertTest = [[UIAlertView alloc] initWithTitle:@"Hello" message:@"More more more" delegate:nil cancelButtonTitle:@"cancel" otherButtonTitles: nil];
            [alertTest show];
            
            [cell hideUtilityButtonsAnimated:YES];
            break;
        }
        case 1:
        {
            // Delete button was pressed
            NSIndexPath *cellIndexPath = [self.tableView indexPathForCell:cell];
            
            [_testArray[cellIndexPath.section] removeObjectAtIndex:cellIndexPath.row];
            [self.tableView deleteRowsAtIndexPaths:@[cellIndexPath] withRowAnimation:UITableViewRowAnimationLeft];
            break;
        }
        default:
            break;
    }
}

- (BOOL)swipeableTableViewCellShouldHideUtilityButtonsOnSwipe:(SWTableViewCell *)cell
{
    // allow just one cell's utility button to be open at once
    return YES;
}

- (BOOL)swipeableTableViewCell:(SWTableViewCell *)cell canSwipeToState:(SWCellState)state
{
    switch (state) {
        case 1:
            // set to NO to disable all left utility buttons appearing
            return YES;
            break;
        case 2:
            // set to NO to disable all right utility buttons appearing
            return YES;
            break;
        default:
            break;
    }
    
    return YES;
}


@end
